app/bundles/CampaignBundle/Tests/Helper/InactiveHelperTest.php

```diff
@@ -0,0 +1,187 @@
+            ->getMock();
+
+        $this->leadRepository = $this->getMockBuilder(LeadRepository::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $this->logger = $this->getMockBuilder(LoggerInterface::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $this->inactiveHelper = new InactiveHelper(
+            $this->scheduler,
+            $this->inactiveContactFinder,
+            $this->eventLogRepository,
+            $this->eventRepository,
+            $this->leadRepository,
+            $this->logger
+        );
+    }
+
+    public function testRemoveContactsThatAreNotApplicable(): void
+    {
+        $lastActiveEventId = 6;
+
+        // lead not applicable because of parent negative path taken
+        $leadNegative = new Lead();
+        $leadNegative->setId(9);
+
+        // lead not applicable because of parent positive path taken
+        $leadNegative2 = new Lead();
+        $leadNegative2->setId(10);
+
+        // applicable lead
+        $leadPositive = new Lead();
+        $leadPositive->setId(12);
+
+        // lead not applicable because of no parent event log
+        $leadNegative3 = new Lead();
+        $leadNegative3->setId(11);
+
+        $this->eventLogRepository->expects($this->once())
+            ->method('getDatesExecuted')
+            ->willReturn(new ArrayCollection([
+                $leadNegative->getId()  => DateTime::createFromFormat('Y-m-d H:i:s', '2022-05-28 21:37:00'),
+                $leadNegative2->getId() => DateTime::createFromFormat('Y-m-d H:i:s', '2022-05-28 21:37:00'),
+                $leadPositive->getId()  => DateTime::createFromFormat('Y-m-d H:i:s', '2022-05-28 21:37:00'),
+                $leadNegative3->getId() => DateTime::createFromFormat('Y-m-d H:i:s', '2022-05-28 21:37:00'),
+            ]));
+
+        $log = $this->getMockBuilder(LeadEventLog::class)
+            ->getMock();
+        $log->expects($this->exactly(3))
+            ->method('getNonActionPathTaken')
+            ->will($this->onConsecutiveCalls(1, 0, 1));
+
+        $campaign = $this->getMockBuilder(Campaign::class)
+            ->getMock();
+        $campaign->expects($this->any())
+            ->method('getId')
+            ->willReturn(2);
+
+        $parentEvent = $this->getMockBuilder(Event::class)
+            ->getMock();
+        $parentEvent->expects($this->exactly(4))
+            ->method('getLogByContactAndRotation')
+            ->will($this->onConsecutiveCalls($log, $log, $log, null));
+
+        $event = $this->getMockBuilder(Event::class)
+            ->getMock();
+        $event->expects($this->once())
+            ->method('getParent')
+            ->willReturn($parentEvent);
+        $event->expects($this->exactly(4))
+            ->method('getDecisionPath')
+            ->willReturn('yes');
+        $event->expects($this->exactly(4))
+            ->method('getCampaign')
+            ->willReturn($campaign);
```

Comment:
We should avoid mocks wherever possible to test code closer to reality. It would be great if you find a way how to avoid other mocks in this test. I don't want to force you to more work. I really like this test! Just in case you want it. There are many articles if you want to dig deeper. Like [this one](https://tomasvotruba.com/blog/2018/06/11/how-to-turn-mocks-from-nightmare-to-solid-kiss-tests/).
```suggestion
        $event = new Event();
        $event->setParent($parentEvent);
        $event->setDecisionPath('yes');
        $event->setCampaign($campaign);
```